cpumanager: expand test coverage for prefer-align-cpus-by-uncore-cache

wongchar opened the pull request:
#### What type of PR is this?
/kind cleanup 

#### What this PR does / why we need it:
Expands test coverage for `prefer-align=cpus-by-uncore-cache` CPUManager policy option to satisfy beta graduation requirements

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

<https://github.com/kubernetes/enhancements/issues/5109>

#### Special notes for your reviewer:

Added test coverage includes:
- Unit tests and e2e tests for `prefer-align-cpus-by-uncore-cache` and `strict-cpu-reservation` compatibility
- Unit tests for `prefer-align-cpus-by-uncore-cache` feature incompatibility with ` distribute-cpus-across-numa` and `distribute-cpus-across-cores`
- e2e tests for `prefer-align-cpus-by-uncore-cache` disabled

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

wongchar committed:
expand test coverage for uncore alignment

add feature compatibility

check uncore cpuset alignment

check shared uncores

ffromani reviewed (commented):
Yes, we need this test coverage extension.
minorish questions inside, PTAL. Other than that I think we can merge once questions addressed

SergeyKanzhelev commented:
/assign @ffromani

since you already looking into this

ffromani commented:
> > Ensuring the feature can be safely disabled is a requirement for beta graduation, because the feature is on by default.
> 
> @ffromani just to clarify, my assumption was that this feature is disabled by default (I may need to deliberate internally with my organization). Uncore cache alignment as an optimization is use-case specific, there are applications that won't benefit, primarily applications dependent on memory-bandwidth. My opinion is that it should be opt-in

I'm expressed myself poorly, let me rephrase. When we add features in kubernetes, we guard them with feature gates (FGs) till they become GA. When a feature is alpha, it is disabled by default, and opt-in. When it becomes beta, it becomes enabled by default, so it is opt-out. Then, as part as the beta graduation, we need to demonstrate the feature can actually be safely disabled reverting to the old behavior (if any) or to a stable state. This is the general rule.

In this case the FG guards the availability of the feature which is always opt-in. With the FG enable, users can set the option, and then they have top actually set the option.

1. FG=false -> no LLC alignment possible at all, option disabled
2. FG=true, option not set -> LLC alignment ignored
3. FG=true, option set -> LLC alignment done in a best-effort way

Still, despite the peculiarity of this case (double opt-in) we need a test to comply with the general case: if the FG is disabled, no LLC alignment is ever attempted because the option is short-circuited to off.
IOW, we need to ensure that the code we added can be completely skipped. We had bugs in the past in kubernetes because the new code was not completely guarded by its FG.

Perhaps this test already exist, I don't remember; if not, it should be easy to add, it's sufficient to ensure that the current code still performs exclusive CPU allocation.

In a nutshell, there is a component of general due diligence we need to do, even if this specific case carries less risk than usual.

I hope I expressed myself better now!

swatisehgal commented on test/e2e_node/cpumanager_test.go:
```diff
@@ -1080,6 +1136,103 @@ var _ = SIGDescribe("CPU Manager", ginkgo.Ordered, ginkgo.ContinueOnFailure, fra
+						siblingsCPUs := makeThreadSiblingCPUSet(cpus)
+						gomega.Expect(pod).To(HaveContainerCPUsEqualTo(cnt.Name, siblingsCPUs))
+
+						// expect full uncore cache worth of cpus to be assigned to uncoreCacheID equal to 1
+						// since CPU0 is part of reserved CPUset, resulting in insufficient CPUs from
+						// uncoreCacheID equal to 0
+						expUncoreCPUSet, err := uncoreCPUSetFromSysFS(1)
```
Sounds good to me, Thanks!

ffromani commented on test/e2e_node/cpumanager_test.go:
```diff
@@ -2142,63 +2168,22 @@ func cpuSiblingListFromSysFS(cpuID int64) cpuset.CPUSet {
-	result := cpuset.New()
-	entries, err := os.ReadDir(basePath)
-	// return error if base path directory does not exist
+func uncoreCacheIDFromSysFS(cpuID int) (int64, error) {
```
perhaps this can be simplified, but it's an optimization we can totally do later.

wongchar replied:
kept this the same for now

ffromani replied:
fine for me

ffromani commented on test/e2e_node/cpumanager_test.go:
```diff
@@ -1102,12 +1098,8 @@ var _ = SIGDescribe("CPU Manager", ginkgo.Ordered, ginkgo.ContinueOnFailure, fra
-						framework.ExpectNoError(err, "cannot determine shared cpus for uncore cache on node")
-						gomega.Expect(pod).To(HaveContainerCPUsEqualTo(cnt.Name, expUncoreCPUSet))
+						// expect full uncore cache worth of cpus to be assigned from the same uncore cache block
+						gomega.Expect(pod).To(HaveContainerCPUsWithSameUncoreCacheID(cnt.Name))
```
ok, we need also to check this uncore ID is different from the uncore ID the reserved CPUs are affine to, because this was previously captured by the test.

ffromani replied:
Or not? We totally do if we have strict-cpu-reservation set, that's fine. And this was added and it works. But in this case I don't remember if we want to enforce this.

ffromani reviewed (commented):
Thanks for the updates. IMO we are almost there. What's needed is:
1. we agreed about one single missing FG disablement test, let's add it. Should be not too hard
2. the changes to check all cores are affine to the same Uncore ID reegardless of the exact ID looks fine; let's add the missing part: in a couple spots we need to check all cores are affine to the same uncore ID, but that uncore ID is different from the uncore ID another set of cpus (namely the reserved ones) are affine to.

I added suggestions inside to implement that, should be reasonnable.

Once the above is done, we can merge and to the best of my knowledge that's the missing part to move to beta (then we have docs to fix, but that's past code freeze).

wongchar commented:
1. Added feature-gate disabled test. the `enableCPUManagerOptions` bool is used for enabling alpha and beta. I assumed this was ok, but I can break them up if preferred and update all tests.
2. Added checks to confirm if container CPUSet has any shared uncore caches with referenced CPUSet
3. Squashed commits

ffromani reviewed (commented):
thanks @wongchar , my comments are addressed

ffromani commented:
/lgtm

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-topology-manager

swatisehgal commented:
All my comments have been addressed as well, once the tests are passing in the CI we can move ahead with this PR.

ffromani commented on test/e2e_node/cpumanager_test.go:
```diff
@@ -813,10 +1088,8 @@ var _ = SIGDescribe("CPU Manager", ginkgo.Ordered, ginkgo.ContinueOnFailure, fra
-					framework.ExpectNoError(err, "cannot determine shared cpus for uncore cache on node")
-					gomega.Expect(pod).To(HaveContainerCPUsASubsetOf(cnt.Name, expUncoreCPUSet))
+					gomega.Expect(pod).To(HaveContainerCPUsWithSameUncoreCacheID(cnt.Name))
+					gomega.Expect(pod).ToNot(HaveContainerCPUsShareUncoreCacheWith(cnt.Name, reservedCPUs))
```
I think we can remove this very line, which is causing the failure. We don't guarantee this behavior (or try hard to ensure).
I think my previous comment was misplaced indeed. This specific test should not check that the allocated uncore is shared, we should do only if `strict-cpu-reservation` is set. And we do that also for `full-pcpus-only` (lines 1376 and 1465) and this works as expected.

ffromani commented on test/e2e_node/cpumanager_test.go:
```diff
@@ -1798,6 +2211,87 @@ func HaveContainerCPUsQuasiThreadSiblings(ctnName string, toleration int) types.
+	md := &msgData{
+		Name: ctnName,
+	}
+	return gcustom.MakeMatcher(func(actual *v1.Pod) (bool, error) {
+		containerCPUs, err := getContainerAllowedCPUs(actual, ctnName, false)
+		if err != nil {
+			return false, fmt.Errorf("getContainerAllowedCPUs(%s) failed: %w", ctnName, err)
+		}
+		md.CurrentCPUs = containerCPUs.String()
+
+		// Build set of uncore cache IDs from the reference cpuset
+		refUncoreIDs := make(map[int64]struct{})
+		for _, cpu := range ref.List() {
+			uncoreID, err := uncoreCacheIDFromSysFS(cpu)
+			if err != nil {
+				return false, fmt.Errorf("failed to read uncore cache ID for reference CPU %d: %w", cpu, err)
+			}
+			refUncoreIDs[uncoreID] = struct{}{}
+		}
+
+		// Check if any container CPUs share an uncore ID with the reference set
+		for _, cpu := range containerCPUs.List() {
+			uncoreID, err := uncoreCacheIDFromSysFS(cpu)
+			if err != nil {
+				return false, fmt.Errorf("failed to read uncore cache ID for container CPU %d: %w", cpu, err)
+			}
+			if _, ok := refUncoreIDs[uncoreID]; ok {
+				md.UncoreCacheAlign = fmt.Sprintf("container CPU %d shares uncore ID %d with reference cpuset", cpu, uncoreID)
+				return true, nil
+			}
+		}
+
+		// No shared uncore IDs found
+		md.UncoreCacheAlign = fmt.Sprintf("container %s CPUs do not share any uncore cache ID with reference cpuset %s", ctnName, ref.String())
+		return false, nil
+	}).WithTemplate(
+		"Pod {{.Actual.Namespace}}/{{.Actual.Name}} UID {{.Actual.UID}} container {{.Data.Name}} has CPUSet <{{.Data.CurrentCPUs}}> with no shared uncore cache ID with reference CPUSet",
+		md,
+	)
+}
```
if you resubmit, suggested improvement:
```
@@ -2147,7 +2148,8 @@ func HaveContainerCPUsWithSameUncoreCacheID(ctnName string) types.GomegaMatcher
 
 func HaveContainerCPUsShareUncoreCacheWith(ctnName string, ref cpuset.CPUSet) types.GomegaMatcher {
        md := &msgData{
-               Name: ctnName,
+               Name:         ctnName,
+               ExpectedCPUs: ref.String(),
        }
        return gcustom.MakeMatcher(func(actual *v1.Pod) (bool, error) {
                containerCPUs, err := getContainerAllowedCPUs(actual, ctnName, false)
@@ -2157,32 +2159,30 @@ func HaveContainerCPUsShareUncoreCacheWith(ctnName string, ref cpuset.CPUSet) ty
                md.CurrentCPUs = containerCPUs.String()
 
                // Build set of uncore cache IDs from the reference cpuset
-               refUncoreIDs := make(map[int64]struct{})
-               for _, cpu := range ref.List() {
+               refUncoreIDs := sets.New[int64]()
+               for _, cpu := range ref.UnsortedList() {
                        uncoreID, err := uncoreCacheIDFromSysFS(cpu)
                        if err != nil {
                                return false, fmt.Errorf("failed to read uncore cache ID for reference CPU %d: %w", cpu, err)
                        }
-                       refUncoreIDs[uncoreID] = struct{}{}
+                       refUncoreIDs.Insert(uncoreID)
                }
 
                // Check if any container CPUs share an uncore ID with the reference set
-               for _, cpu := range containerCPUs.List() {
+               for _, cpu := range containerCPUs.UnsortedList() {
                        uncoreID, err := uncoreCacheIDFromSysFS(cpu)
                        if err != nil {
                                return false, fmt.Errorf("failed to read uncore cache ID for container CPU %d: %w", cpu, err)
                        }
-                       if _, ok := refUncoreIDs[uncoreID]; ok {
-                               md.UncoreCacheAlign = fmt.Sprintf("container CPU %d shares uncore ID %d with reference cpuset", cpu, uncoreID)
+                       if refUncoreIDs.Has(uncoreID) {
+                               md.UncoreCacheAlign = fmt.Sprintf("%d", uncoreID)
                                return true, nil
                        }
                }
 
-               // No shared uncore IDs found
-               md.UncoreCacheAlign = fmt.Sprintf("container %s CPUs do not share any uncore cache ID with reference cpuset %s", ctnName, ref.String())
                return false, nil
        }).WithTemplate(
-               "Pod {{.Actual.Namespace}}/{{.Actual.Name}} UID {{.Actual.UID}} container {{.Data.Name}} has CPUSet <{{.Data.CurrentCPUs}}> with no shared uncore cache ID with reference CPUSet",
+               "Pod {{.Actual.Namespace}}/{{.Actual.Name}} UID {{.Actual.UID}} container {{.Data.Name}} has CPUSet <{{.Data.CurrentCPUs}}> sharing uncoreCache ID <{{.Data.UncoreCacheAlign}}> with reference CPUSet <{{.Data.ExpectedCPUs}}>",
                md,
        )
 }
```

ffromani replied:
note this require to add the import (which is fine):
```
k8s.io/apimachinery/pkg/util/sets
```

wongchar committed:
fix uncore e2e check

wongchar commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

wongchar commented:
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani reviewed (commented):
/approve
/lgtm

all comments addressed, relevant lanes are green
